torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -973,18 +1868,26 @@ void SegmentCandidateFinder::findSegments() {
   // Need this for initialization of the DAG that is process
   std::unordered_map<Expr*, SegmentedGroup*> expr2group;
 
+  // Keep track of complete fusion input use
+  std::unordered_map<Val*, SegmentedGroup*> input2group;
+
   // Initialize DAG, convert each expr to a segment group
-  size_t total_tv_exprs = 0;
   auto exprs = completeFusion().exprs();
   for (auto expr : exprs) {
     if (!ir_utils::isScalarOp(expr)) {
       auto new_group = segmented_fusion_->newGroup(expr);
       expr2group.insert(std::make_pair(expr, new_group));
-      total_tv_exprs++;
     }
   }
 
-  segmented_fusion_->total_tv_expr_count_ = total_tv_exprs;
+  // Insert auxiliary groups to use group dependency on inputs as well
+  for (auto input : completeFusion().inputs()) {
+    // These groups are used to represent input as a common
+    //  producer in horizontal merges, and should never be
+    //  seen as a candidate for vertical merge
+    auto new_group = segmented_fusion_->newGroup();
+    input2group.insert({input, new_group});
+  }
```

Comment:
These groups won't be used in vertical merges, but it goes a bit beyond that in that they will never be merged into a segmented group. I think that latter property is pretty important to mention. If they do start getting merged in I don't know what the side effects would be, but doesn't seem good. This is the one thing I don't like about this design that I think we should try to refactor out. I understand why you did it, but I don't like having groups that are intended to wrap expressions also wrapping just input values. Can mark as a TODO: for a later refactor.